eth/beacon/block_committees_info.py

```diff
@@ -0,0 +1,20 @@
+from typing import (
+    Iterable,
+    NamedTuple,
+    TYPE_CHECKING,
+)
+
+if TYPE_CHECKING:
+    from eth.beacon.types.shard_and_committees import ShardAndCommittee  # noqa: F401
+
+
+BlockCommitteesInfo = NamedTuple(
+    'BlockCommitteesInfo',
+    (
+        ('proposer_index', int),
+        ('proposer_index_in_committee', int),
+        ('proposer_shard_id', int),
+        ('proposer_committee_size', int),
+        ('shards_and_committees', Iterable['ShardAndCommittee'])
```

Comment:
Iterable is probably not what you want since that would allow a generator.  I'd actually recommend being very specific here and choosing `Tuple`.

Reply:
:+1: to what @pipermerriam said.

Taking the chance to elaborate on my comment here: https://github.com/ethereum/py-evm/pull/1373#discussion_r224106008

I think as a rule of thumb: APIs *taking parmeters* (e.g. `def do_something(things: Iterable[Foo])` should accept the most flexible interface that the API can work with. I can hardly think of a reason why an API should constrain a parameter to a concrete type (e.g. `Tuple[Foo]`) if there is nothing intrinsically preventing that API from working with an `Iterable[Foo]`.

However, **Return types** should *usually* be concrete. Because, most of the time an API returning *something*, does have intrinsic knowledge about the type anyway and there is nothing to gain from making the return type more abstract if that return type fulfills that abstract interface anyway.

E.g. There is nothing to gain from doing this:

```python
def get_foo() -> Iterable[Foo]
    return [foo, too]
```

Even though that works, we are not improving flexibility here. If we make the return type `List[Foo]`, that `List[Foo]` still fulfills being `Iterable[Foo]` as well as `Sequence[Foo]` implicitly, so by making the return type somthing like `Iterable[Foo]` when in reality we have a `List[Foo]` is in fact only limiting the flexibility here.

Coming back to the code in question `shards_and_committees` is, after all, just a property on `BlocksCommitteesInfo`, hence making it fall in the return type bucket.

Related: [Covariance and Contravariance](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science))

Reply:
Also, here is an alternative syntax, if you like it better:

```py
class BlockCommitteesInfo(NamedTuple):
  proposer_index: int
  proposer_index_in_committee: int
  # etc...
```

Reply:
@carver I *think* that syntax does not work with Python 3.5 though

Reply:
I'd be in support to drop 3.5 at any time 🤭😅

Reply:
Two things,

- Can't drop 3.5 until we do it in web3.py  (is v5 py36+?)
- @cburgdorf would you mind migrating that writup to https://github.com/pipermerriam/ethereum-dev-tactical-manual, maybe under a new section for static type hinting.

Reply:
#1391 for discussion

Reply:
@pipermerriam proposed section on typing (choosing types for function parameters and return values)! https://github.com/pipermerriam/ethereum-dev-tactical-manual/pull/24